Pattern changes caused by commit: 72dcc298d335721c053444249c157e9a6431ebea

From: Decorator-2
To:   Decorator-1

From: Flyweight-2
To:   Flyweight-1

From: Mediator-3
To:   Mediator-1

From: Strategy-0
To:   Strategy-1


=========================
       NEW GIT LOG
=========================

This commit refers to file: VALID-7701.txt 

commit 72dcc298d335721c053444249c157e9a6431ebea
Author: Jonathan Ellis <jbellis@apache.org>

    fix indexing empty column values
    patch by jbellis for CASSANDRA-4832



==================================
 Issue CASSANDRA-4832 Description 
=======================================

Project: Cassandra
-----------------

-----------------
Title: [CASSANDRA-4832] AssertionError: keys must not be empty
-----------------

-----------------
Summary: AssertionError: keys must not be empty
-----------------

-----------------
Issue type: Bug
-----------------

-----------------
Current status: Resolved
-----------------

-----------------
Created at: Thu, 18 Oct 2012 05:15:28 +0000
-----------------

-----------------
Resolved at: Mon, 22 Oct 2012 16:50:31 +0000
-----------------

-----------------
Assigned to: Tristan Seligmann
-----------------

-----------------
Description: 

I'm getting errors like this logged:

 INFO 07:08:32,104 Compacting <span
class="error">&#91;SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hf-114-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hf-113-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hf-110-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hd-108-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hd-106-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hd-107-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hf-112-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hf-109-Data.db&#39;),
SSTableReader(path=&#39;/var/lib/cassandra/data/Fusion/quoteinfo/Fusion-quoteinfo.quoteinfo_search_value_idx-hf-111-Data.db&#39;)&#93;</span><br/>ERROR
07:08:32,108 Exception in thread Thread<span
class="error">&#91;CompactionExecutor:5,1,main&#93;</span><br/>java.lang.AssertionError:
Keys must not be empty<br/>        at
org.apache.cassandra.io.sstable.SSTableWriter.beforeAppend(SSTableWriter.java:133)<br/>   
    at org.apache.cassandra.io.sstable.SSTableWriter.append(SSTableWriter.java:154)<br/>  
     at
org.apache.cassandra.db.compaction.CompactionTask.execute(CompactionTask.java:159)<br/>   
    at
org.apache.cassandra.db.compaction.CompactionManager$1.runMayThrow(CompactionManager.java:154)<br/>
       at org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)<br/>    
   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)<br/>       
at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)<br/>        at
java.util.concurrent.FutureTask.run(FutureTask.java:138)<br/>        at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)<br/>  
     at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)<br/>      
 at java.lang.Thread.run(Thread.java:662)

I'm not really sure when this started
happening; they tend to be logged during a repair but I can't reproduce the error 100%
reliably.
 

-----------------

-----------------
Comments: 

New Comment: 
Actually, it seems like this gets logged any time Cassandra tries to flush this column
family now. 


New Comment: 
After some further investigation, it looks like this began after I upgraded from 1.1.2 to
1.1.6. The assertion seems to have been introduced as part of some debugging in <a
href="https://issues.apache.org/jira/browse/CASSANDRA-4687" title="Exception:
DecoratedKey(xxx, yyy) != DecoratedKey(zzz, kkk)" class="issue-link"
data-issue-key="CASSANDRA-4687"><del>CASSANDRA-4687</del></a>. I'm not familiar with the
code at all, but it seems to me that perhaps the assertion is bogus? If I have a secondary
index on a particular column, and that column has an empty value in some row, would this
not result in an empty key in the secondary index column family? 


New Comment: 
I modified my application to avoid inserting the empty column in question (which was a bug
anyway; if that particular column was empty, the application should not have been doing an
insert at all) and the problem has gone away, so it seems like my hypothesis is at least
partially correct. I'm going to attempt to reproduce this in a more controlled environment
now that my immediate issue in production has been resolved. 


New Comment: 
you're right that the assert is bogus.  removed it in
72dcc298d335721c053444249c157e9a6431ebea. 


New Comment: 
Came across this investigating an apparent deadlock in Schema Migrations.If this assertion
fails on the flushWriter executor, it blocks indefinitely. Anything upstream locking-wise
gets stuck also. This was on 1.1.6.Log output below, thread dump attached.ERROR <span
class="error">&#91;FlushWriter:3&#93;</span> 2012-10-19 22:27:56,948
org.apache.cassandra.service.AbstractCassandraDaemon Exception in thread Thread<span
class="error">&#91;FlushWriter:3,5,main&#93;</span><br/>java.lang.AssertionError: Keys
must not be empty<br/>        at
org.apache.cassandra.io.sstable.SSTableWriter.beforeAppend(SSTableWriter.java:133)<br/>   
    at org.apache.cassandra.io.sstable.SSTableWriter.append(SSTableWriter.java:176)<br/>  
     at org.apache.cassandra.db.Memtable.writeSortedContents(Memtable.java:295)<br/>      
 at org.apache.cassandra.db.Memtable.access$600(Memtable.java:48)<br/>        at
org.apache.cassandra.db.Memtable$5.runMayThrow(Memtable.java:316)<br/>        at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)<br/>        at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)<br/>  
     at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)<br/>      
 at java.lang.Thread.run(Thread.java:662) 


New Comment: 
I've run into the same symptoms as Chris, but while attempting to migrate data between a
1.1.1 cluster and 1.1.6.Whilst using sstableloader to import the data, nodes would hit
certain column indexes and stop processing further requests once the assertion was thrown
- restarting the node would almost immediately throw the assertion again, and the node
would just fail to rejoin the ring.tpstats would show active flushwriter tasks but no
further node activity.We had to work around the issue by:1. Removing all indexes from our
target cluster schema<br/>2. Importing the data via sstableloader<br/>3. Scanning through
relevant column families and inserting data into any empty indexed columns<br/>4.
Re-applying the indexes to the target cluster schemaOnly then was the migration
successful.I'll also note that attempting to apply an index to a column which has null
data will also throw the cluster out of sync as the nodes which throw the assertion fail
to migrate their schemas properly INFO <span class="error">&#91;Creating index:
Transactions.TransactionsCountryCode&#93;</span> 2012-10-21 07:45:25,978
ColumnFamilyStore.java (line 659) Enqueuing flush of
Memtable-Transactions.TransactionsCountryCode@1802367190(38862/169512 serialized/live
bytes, 762 ops)<br/> INFO <span class="error">&#91;Creating index:
Transactions.TransactionsStatus&#93;</span> 2012-10-21 07:45:25,980 ColumnFamilyStore.java
(line 659) Enqueuing flush of
Memtable-Transactions.TransactionsStatus@673679943(38862/125966 serialized/live bytes, 762
ops)<br/> INFO <span class="error">&#91;FlushWriter:1&#93;</span> 2012-10-21 07:45:25,987
Memtable.java (line 264) Writing
Memtable-Transactions.TransactionsCountryCode@1802367190(38862/169512 serialized/live
bytes, 762 ops)<br/> INFO <span class="error">&#91;FlushWriter:2&#93;</span> 2012-10-21
07:45:26,004 Memtable.java (line 264) Writing
Memtable-Transactions.TransactionsStatus@673679943(38862/125966 serialized/live bytes, 762
ops)<br/>ERROR <span class="error">&#91;FlushWriter:1&#93;</span> 2012-10-21 07:45:26,004
AbstractCassandraDaemon.java (line 135) Exception in thread Thread<span
class="error">&#91;FlushWriter:1,5,main&#93;</span><br/>java.lang.AssertionError: Keys
must not be empty<br/>	at
org.apache.cassandra.io.sstable.SSTableWriter.beforeAppend(SSTableWriter.java:133)<br/>	at
org.apache.cassandra.io.sstable.SSTableWriter.append(SSTableWriter.java:176)<br/>	at
org.apache.cassandra.db.Memtable.writeSortedContents(Memtable.java:295)<br/>	at
org.apache.cassandra.db.Memtable.access$600(Memtable.java:48)<br/>	at
org.apache.cassandra.db.Memtable$5.runMayThrow(Memtable.java:316)<br/>	at
org.apache.cassandra.utils.WrappedRunnable.run(WrappedRunnable.java:30)<br/>	at
java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)<br/>	at
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)<br/>	at
java.lang.Thread.run(Thread.java:662)<br/> INFO <span
class="error">&#91;FlushWriter:2&#93;</span> 2012-10-21 07:45:26,049 Memtable.java (line
305) Completed flushing
/var/lib/cassandra/data/xxxxxx/Transactions/xxxxxx-Transactions.TransactionsStatus-hf-2-Data.db
(35259 bytes) for commitlog position ReplayPosition(segmentId=1350805525722,
position=0)<br/> INFO <span class="error">&#91;Creating index:
Transactions.TransactionsLastUpdateDate&#93;</span> 2012-10-21 07:45:26,313
ColumnFamilyStore.java (line 659) Enqueuing flush of
Memtable-Transactions.TransactionsLastUpdateDate@1912098049(38862/240277 serialized/live
bytes, 762 ops)<br/> INFO <span class="error">&#91;FlushWriter:2&#93;</span> 2012-10-21
07:45:26,314 Memtable.java (line 264) Writing
Memtable-Transactions.TransactionsLastUpdateDate@1912098049(38862/240277 serialized/live
bytes, 762 ops)<br/> INFO <span class="error">&#91;FlushWriter:2&#93;</span> 2012-10-21
07:45:26,743 Memtable.java (line 305) Completed flushing
/var/lib/cassandra/data/xxxxxx/Transactions/xxxxxx-Transactions.TransactionsLastUpdateDate-hf-2-Data.db
(37024 bytes) for commitlog position ReplayPosition(segmentId=1350805525722,
position=0)<br/> INFO <span class="error">&#91;main&#93;</span> 2012-10-21 07:45:27,052
CommitLogReplayer.java (line 272) Finished reading
/var/lib/cassandra/commitlog/CommitLog-1350768744805.log<br/> INFO <span
class="error">&#91;main&#93;</span> 2012-10-21 07:45:27,054 ColumnFamilyStore.java (line
659) Enqueuing flush of Memtable-Versions@1851630436(83/103 serialized/live bytes, 3
ops)<br/> INFO <span class="error">&#91;FlushWriter:2&#93;</span> 2012-10-21 07:45:27,054
Memtable.java (line 264) Writing Memtable-Versions@1851630436(83/103 serialized/live
bytes, 3 ops)<br/> INFO <span class="error">&#91;FlushWriter:2&#93;</span> 2012-10-21
07:45:27,061 Memtable.java (line 305) Completed flushing
/var/lib/cassandra/data/system/Versions/system-Versions-hf-1-Data.db (247 bytes) for
commitlog position ReplayPosition(segmentId=1350805525722, position=0) 


New Comment: 
Is there an actual patch for this, or a Workaround? We see this same AssertionError when
attempting to start our nodes; restarting the node usually works correctly, however. 


